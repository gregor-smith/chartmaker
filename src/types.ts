import {
    Record as Record_,
    Literal,
    String,
    Union,
    InstanceOf,
    Boolean,
    Static,
    Number as Number_,
    Array as Array_,
    Runtype,
    Partial as Partial_
} from 'runtypes'

import {
    CHART_ALBUMS_COUNT,
    MAX_SCREENSHOT_SCALE,
    MAX_COLLAGE_ROWS_X,
    MAX_COLLAGE_ROWS_Y,
    STATE_VERSION,
    EXPORT_CHART_PLACEHOLDER
} from '@/constants'


const Integer = Number_.withConstraint(Number.isSafeInteger)


function IntegerRange(minimum: number, maximum = Number.MAX_SAFE_INTEGER) {
    return Integer.withConstraint(number => number >= minimum && number <= maximum)
}


function FixedSizeArray<T extends Runtype>(element: T, size: number) {
    return Array_(element).withConstraint(array => array.length === size)
}


function NonEmptyArray<T extends Runtype>(element: T) {
    return Array_(element).withConstraint(array => array.length > 0)
}


const NonEmptyString = String.withConstraint(string => string.length !== 0)


const PositiveInteger = IntegerRange(0)


const V1PlaceholderAlbum = Record_({
    placeholder: Literal(true),
    id: PositiveInteger
})


const SearchAlbum = Record_({
    name: NonEmptyString,
    url: NonEmptyString
})


const V1NamedAlbum = SearchAlbum.And(
    Record_({
        placeholder: Literal(false),
        id: PositiveInteger
    })
)
const V2NamedAlbum = SearchAlbum.And(
    Record_({
        id: PositiveInteger
    })
)
export const NamedAlbum = V2NamedAlbum


const V1Album = V1NamedAlbum.Or(V1PlaceholderAlbum)
const V2Album = V2NamedAlbum.Or(PositiveInteger)
const Album = V2Album


const ChartShape = Union(
    Record_({
        tag: Literal('Top'),
        size: Union(
            Literal(40),
            Literal(42),
            Literal(100)
        )
    }),
    Record_({
        tag: Literal('Collage')
    })
)


const V1Chart = Record_({
    name: NonEmptyString,
    albums: FixedSizeArray(V1Album, CHART_ALBUMS_COUNT),
    shape: ChartShape,
    rowsX: IntegerRange(1, MAX_COLLAGE_ROWS_X),
    rowsY: IntegerRange(1, MAX_COLLAGE_ROWS_Y)
})
const V2Chart = Record_({
    ...V1Chart.fields,
    albums: FixedSizeArray(Album, CHART_ALBUMS_COUNT)
})
const Chart = V2Chart


const SearchState = Union(
    Record_({
        tag: Literal('Waiting')
    }),
    Record_({
        tag: Literal('Loading'),
        controller: InstanceOf(AbortController)
    }),
    Record_({
        tag: Literal('Complete'),
        albums: NonEmptyArray(SearchAlbum)
    }),
    Record_({
        tag: Literal('Error'),
        message: String
    })
).And(
    Record_({
        query: String
    })
)


const ScreenshotState = Record_({
    loading: Boolean,
    scale: IntegerRange(1, MAX_SCREENSHOT_SCALE)
})


const ExportChartAlbum = SearchAlbum.Or(Literal(EXPORT_CHART_PLACEHOLDER))

const ExportChartShape = ChartShape.alternatives[0].Or(
    ChartShape.alternatives[1].And(
        Record_({
            rowsX: V2Chart.fields.rowsX,
            rowsY: V2Chart.fields.rowsY,
        })
    )
)

const ExportChart = Record_({
    name: V2Chart.fields.name,
    albums: FixedSizeArray(ExportChartAlbum, CHART_ALBUMS_COUNT),
    shape: ExportChartShape
})


export const V1State = Record_({
    apiKey: String,
    charts: NonEmptyArray(V1Chart),
    activeChartIndex: PositiveInteger,
    search: SearchState,
    screenshot: ScreenshotState
}).And(
    Partial_({
        version: Literal(1),
        highlightedID: PositiveInteger
    })
)
export const V2State = Record_({
    ...V1State.intersectees[0].fields,
    version: Literal(2),
    charts: NonEmptyArray(V2Chart)
}).And(
    Partial_({
        highlightedID: PositiveInteger
    })
)
const V3State = Record_({
    ...V2State.intersectees[0].fields,
    version: Literal(STATE_VERSION)
}).And(
    Partial_({
        ...V2State.intersectees[1].fields,
        viewing: ExportChart
    })
)
export const State = V3State


export type NamedAlbum = Static<typeof NamedAlbum>
export type SearchAlbum = Static<typeof SearchAlbum>
export type Album = Static<typeof Album>
export type ChartShape = Static<typeof ChartShape>
export type Chart = Static<typeof Chart>
export type SearchState = Static<typeof SearchState>
export type ScreenshotState = Static<typeof ScreenshotState>
export type State = Static<typeof State>
export type V1State = Static<typeof V1State>
export type V2State = Static<typeof V2State>
export type ExportChart = Static<typeof ExportChart>
export type ExportChartShape = Static<typeof ExportChartShape>
export type ExportChartAlbum = Static<typeof ExportChartAlbum>
