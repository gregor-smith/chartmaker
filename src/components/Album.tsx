import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
