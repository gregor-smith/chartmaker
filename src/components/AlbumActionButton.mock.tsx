const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
