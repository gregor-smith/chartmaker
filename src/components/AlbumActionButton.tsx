import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
