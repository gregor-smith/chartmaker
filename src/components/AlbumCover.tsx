import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
