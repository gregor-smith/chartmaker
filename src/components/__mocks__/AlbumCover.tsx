import type { FC } from 'react'
import { cx } from 'emotion'

import type { AlbumCoverProps } from '@/components/AlbumCover'


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    overlayClass,
    children,
    highlighted,
    ...props
}) => {
    const json = JSON.stringify(album)
    const className = cx('mock-album-cover', overlayClass)
    return (
        <div {...props} className={className}>
            {`Album: ${json}`}
            {`Size: ${size}`}
            {`Highlighted: ${highlighted}`}
            {children}
        </div>
    )
}
