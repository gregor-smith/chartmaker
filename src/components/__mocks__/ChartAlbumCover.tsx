import type { FC } from 'react'

import type { ChartAlbumCoverProps } from '@/components/ChartAlbumCover'


export const ChartAlbumCover: FC<ChartAlbumCoverProps> = ({ album, size, highlighted, children, ...props }) => {
    const json = JSON.stringify(album)
    return (
        <div {...props} className='mock-chart-album-cover'>
            {`Album: ${json}`}
            {`Size: ${size}`}
            {`Highlighted: ${highlighted}`}
            {children}
        </div>
    )
}
