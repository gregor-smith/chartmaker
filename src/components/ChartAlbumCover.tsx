import { h, FunctionComponent, Fragment, VNode } from 'preact'

import { Album } from '../state'
import { AlbumCover } from './AlbumCover'
import { DispatchProps } from '../reducer'
import { css } from 'emotion'
import { AlbumActionButton } from './AlbumActionButton'


const chartPattern = /^chart-(.+)$/
const searchPattern = /^search-(.+)$/


function getAlbumID(
    event: DragEvent,
    pattern: RegExp
): number | null {
    for (const type of event.dataTransfer!.types) {
        const match = type.match(pattern)?.[1]
        const id = Number(match)
        if (!Number.isNaN(id)) {
            return id
        }
    }
    return null
}


function dragOver(event: DragEvent) {
    for (const type of event.dataTransfer!.types) {
        if (chartPattern.test(type)) {
            event.dataTransfer!.dropEffect = 'move'
            event.preventDefault()
            return
        }
        if (searchPattern.test(type)) {
            event.dataTransfer!.dropEffect = 'copy'
            event.preventDefault()
            return
        }
    }
}


const overlayStyle = css({
    display: 'flex',
    justifyContent: 'space-between',
    padding: '0.5rem'
})


type Props = DispatchProps<
    | 'DragChartAlbum'
    | 'DropSearchAlbum'
    | 'PromptToRenameAlbum'
    | 'DeleteAlbum'
> & {
    album: Album
    sizeRem: number
}


export const ChartAlbumCover: FunctionComponent<Props> = ({ dispatch, album, sizeRem }) => {
    function dragStart(event: DragEvent) {
        event.dataTransfer!.setData(`chart-${album.id}`, '')
        event.dataTransfer!.effectAllowed = 'copyMove'
    }

    function dragEnter(event: DragEvent) {
        const sourceID = getAlbumID(event, chartPattern)
        if (sourceID === null) {
            return
        }
        dispatch({
            tag: 'DragChartAlbum',
            sourceID,
            targetID: album.id
        })
        event.preventDefault()
    }

    function drop(event: DragEvent) {
        const sourceID = getAlbumID(event, searchPattern)
        if (sourceID === null) {
            return
        }
        dispatch({
            tag: 'DropSearchAlbum',
            sourceID,
            targetID: album.id
        })
        event.preventDefault()
    }

    let buttons: VNode | undefined
    if (!album.placeholder) {
        function rename() {
            dispatch({
                tag: 'PromptToRenameAlbum',
                id: album.id
            })
        }

        function remove() {
            dispatch({
                tag: 'DeleteAlbum',
                id: album.id
            })
        }

        buttons = (
            <Fragment>
                <AlbumActionButton colour='yellow' onClick={rename} title='Rename'>
                    ‚úèÔ∏è
                </AlbumActionButton>
                <AlbumActionButton colour='red' onClick={remove} title='Delete'>
                    üóëÔ∏è
                </AlbumActionButton>
            </Fragment>
        )
    }


    return (
        <AlbumCover album={album}
                sizeRem={sizeRem}
                onDragStart={dragStart}
                onDragOver={dragOver}
                onDragEnter={dragEnter}
                onDrop={drop}
                overlayclass={overlayStyle}>
            {buttons}
        </AlbumCover>
    )
}
