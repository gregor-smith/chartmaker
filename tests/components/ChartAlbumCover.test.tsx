import React, { FC } from 'react'
import { render } from 'react-dom'
import { cx } from 'emotion'
import { act } from 'react-dom/test-utils'

import { Album, Size } from '@/types'
import { ChartAlbumCover, ChartAlbumCoverProps } from '@/components/ChartAlbumCover'
import { AlbumCoverProps } from '@/components/AlbumCover'

import {
    RenderContainer,
    ignore,
    fireEvent,
    DragEventDataTransferMock
} from '../utils'


const container = new RenderContainer()


const TestAlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    overlayClass,
    children,
    ...props
}) => {
    const json = JSON.stringify(album)
    const className = cx('test-album-cover', overlayClass)
    return (
        <div {...props} className={className}>
            {`Album: ${json}`}
            {`Size: ${size}`}
            {children}
        </div>
    )
}


const placeholderAlbum: Album = {
    placeholder: true,
    id: 123
}

const namedAlbum: Album = {
    placeholder: false,
    id: 456,
    name: 'Test named album',
    url: 'https://test.com'
}


// delete the semicolon and watch as vscode's syntax highlighting dies
type ActionParams = Parameters<ChartAlbumCoverProps['dispatch']>;


test('renders album cover for placeholder album', () => {
    render(
        <ChartAlbumCover dispatch={ignore}
            album={placeholderAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('renders album cover with overlay buttons for named album', () => {
    render(
        <ChartAlbumCover dispatch={ignore}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('clicking rename button dispatches action', () => {
    const mock = jest.fn<void, ActionParams>()

    render(
        <ChartAlbumCover dispatch={mock}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    act(() => {
        const button = container.element?.querySelector('button[title="Rename"]')
        fireEvent('click', button)
    })

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<ActionParams>({
        tag: 'PromptToRenameAlbum',
        id: namedAlbum.id
    })
})


test('clicking delete button dispatches action', () => {
    const mock = jest.fn<void, ActionParams>()

    render(
        <ChartAlbumCover dispatch={mock}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    act(() => {
        const button = container.element?.querySelector('button[title="Delete"]')
        fireEvent('click', button)
    })

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<ActionParams>({
        tag: 'DeleteAlbum',
        id: namedAlbum.id
    })
})


test('starting drag sets event data', () => {
    render(
        <ChartAlbumCover dispatch={ignore}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const mock = new DragEventDataTransferMock()

    act(() =>
        fireEvent(
            'dragStart',
            container.element?.firstChild,
            { dataTransfer: mock }
        )
    )

    expect(mock.setDataMock).toHaveBeenCalledTimes(1)
    expect(mock.setDataMock).toHaveBeenCalledWith(`chart-${namedAlbum.id}`, '')
    expect(mock.effectAllowedMock).toHaveBeenCalledTimes(1)
    expect(mock.effectAllowedMock).toHaveBeenCalledWith('copyMove')
})


test('chart album drag enter dispatches action', () => {
    const dispatchMock = jest.fn<void, ActionParams>()

    render(
        <ChartAlbumCover dispatch={dispatchMock}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'chart-1' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'dragEnter',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith<ActionParams>({
        tag: 'DragChartAlbum',
        sourceID: 1,
        targetID: namedAlbum.id
    })
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test.each([ 'search-1', 'some-other-type' ])('any other drag enter is ignored', type => {
    const dispatchMock = jest.fn<void, ActionParams>()

    render(
        <ChartAlbumCover dispatch={dispatchMock}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ type ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'dragEnter',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(0)
    expect(preventDefaultMock).toHaveBeenCalledTimes(0)
})


test('dragging chart album over sets move drop effect', () => {
    render(
        <ChartAlbumCover dispatch={ignore}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'chart-1' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() =>
        fireEvent(
            'dragOver',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    )

    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledTimes(1)
    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledWith('move')
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test('dragging search album over sets move copy effect', () => {
    render(
        <ChartAlbumCover dispatch={ignore}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'search-1' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() =>
        fireEvent(
            'dragOver',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    )

    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledTimes(1)
    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledWith('copy')
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test('dragging anything else over does nothing', () => {
    render(
        <ChartAlbumCover dispatch={ignore}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'some-other-type' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() =>
        fireEvent(
            'dragOver',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    )

    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledTimes(0)
    expect(preventDefaultMock).toHaveBeenCalledTimes(0)
})


test('dropping search album dispatches action', () => {
    const dispatchMock = jest.fn<void, ActionParams>()

    render(
        <ChartAlbumCover dispatch={dispatchMock}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'search-1' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'drop',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith<ActionParams>({
        tag: 'DropSearchAlbum',
        sourceID: 1,
        targetID: namedAlbum.id
    })
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test.each([ 'chart-1', 'some-other-type' ])('dropping anything else does nothing', type => {
    const dispatchMock = jest.fn<void, ActionParams>()

    render(
        <ChartAlbumCover dispatch={dispatchMock}
            album={namedAlbum}
            size={'3rem' as Size}
            albumCoverComponent={TestAlbumCover}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ type ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'drop',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(0)
    expect(preventDefaultMock).toHaveBeenCalledTimes(0)
})
