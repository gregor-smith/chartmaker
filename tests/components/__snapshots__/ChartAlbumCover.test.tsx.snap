// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders album cover for placeholder album 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  padding: 0.5rem;
}

<div
  id="test-render-container"
>
  <div
    class="mock-album-cover emotion-0"
  >
    Album: {"placeholder":true,"id":123}
    Size: 3rem
  </div>
</div>
`;

exports[`renders album cover with overlay buttons for named album 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  padding: 0.5rem;
}

<div
  id="test-render-container"
>
  <div
    class="mock-album-cover emotion-0"
  >
    Album: {"placeholder":false,"id":456,"name":"Test named album","url":"https://test.com"}
    Size: 3rem
    <div
      class="mock-rename-album-button"
    >
      ID: 456
    </div>
    <div
      class="mock-delete-album-button"
    >
      ID: 456
    </div>
  </div>
</div>
`;
